libraries/wallet/wallet.cpp

```diff
@@ -759,37 +759,77 @@ class wallet_api_impl
       return true;
    }
 
-   signed_transaction multisig_sign_transaction( signed_transaction tx,
-                                                 const vector<string> &wif_keys, bool broadcast )
+   signed_transaction multisig_sign_transaction( signed_transaction tx, bool broadcast )
    {
-      vector<fc::ecc::private_key> private_keys;
-
-      for ( const string &wif_key : wif_keys )
+      set<public_key_type> pks = _remote_db->get_potential_signatures( tx );
+      flat_set<public_key_type> owned_keys;
+      owned_keys.reserve( pks.size() );
+      std::copy_if( pks.begin(), pks.end(),
+                    std::inserter( owned_keys, owned_keys.end() ),
+                    [this]( const public_key_type &pk ) {
+                       return _keys.find( pk ) != _keys.end();
+                    } );
+
+      set<public_key_type> approving_key_set =
+         _remote_db->get_required_signatures( tx, owned_keys );
+
+      if ( ( ( tx.ref_block_num == 0 && tx.ref_block_prefix == 0 ) ||
+             tx.expiration == fc::time_point_sec() ) &&
+           tx.signatures.empty() )
       {
-         fc::optional<fc::ecc::private_key> optional_private_key = wif_to_key( wif_key );
+         auto dyn_props = get_dynamic_global_properties();
+         tx.set_reference_block( dyn_props.head_block_id );
 
-         if ( !optional_private_key )
+         // first, some bookkeeping, expire old items from
+         // _recently_generated_transactions since transactions include the head
+         // block id, we just need the index for keeping transactions unique
+         // when there are multiple transactions in the same block.  choose a
+         // time period that should be at least one block long, even in the
+         // worst case.  2 minutes ought to be plenty.
+         fc::time_point_sec oldest_transaction_ids_to_track( dyn_props.time -
+                                                             fc::minutes( 2 ) );
+         auto oldest_transaction_record_iter =
+            _recently_generated_transactions.get<timestamp_index>().lower_bound(
+               oldest_transaction_ids_to_track );
+         auto begin_iter =
+            _recently_generated_transactions.get<timestamp_index>().begin();
+         _recently_generated_transactions.get<timestamp_index>().erase(
+            begin_iter, oldest_transaction_record_iter );
+
+         uint32_t expiration_time_offset = 0;
+         for ( ;; )
          {
-            elog( "Invalid private key: ${i}", ( "i", wif_key ) );
-            FC_THROW("");
-         }
+            tx.set_expiration( dyn_props.time +
+                               fc::seconds( 30 + expiration_time_offset ) );
```

Comment:
For a multisig, 30 seconds is not enough. Need to use max allowed value per default.

Also I think the dupe check is not needed here. Multisig will take a while, so by the time someone retries this we will probably have seen many blocks go by.

Reply:
oh yes, fixed~